RNASeq differential expression analysis
=======================================

## Setup

```{r, message=FALSE}
knitr::opts_chunk$set(message=FALSE, warning =FALSE)
library(DESeq2)
library(Rsubread)
library(tidyverse)
theme_set(theme_bw())
library(goseq)
# library(DESeqAnalysis)
```

## Import data

```{r}
# Set up file locations and sample names
dirs = list.dirs('results/04_map',recursive = FALSE)
dirs
samps = gsub('results/04_map/', '', dirs, fixed = TRUE)
samps


# Define the gtf and genome files
gtf = # path/to/gtf
genome = # path/to/fna

# Bring in the sample data and clean it up
samdat = read.csv('data/sample_data.csv', row.names = 1)
samdat

all(samps == rownames(samdat))
```

## Feature Counts

### Setup

```{r}
# Define a function to count features at the gene level
call_fc = function(bamf, gtf, genome, useMF = TRUE){
    fc = Rsubread::featureCounts(bamf, 
                       annot.ext = gtf, isGTFAnnotationFile = TRUE,
                       GTF.featureType = 'gene', 
                       GTF.attrType = 'gene_id',
                       GTF.attrType.extra = c('transcript_id',
                                              'product'),
                       strandSpecific = 2,
                       genome = genome,
                       isPairedEnd = TRUE,
                       requireBothEndsMapped = TRUE,
                       countChimericFragments = FALSE,
                       nthreads = 10,
                       verbose = TRUE,
                       useMetaFeatures = useMF)
    return(fc)
}

# Set up the data structures in advance to save time
gene_counts = matrix(nrow = 33146, ncol = length(samps))
colnames(gene_counts) = samps
```

### Run on the first sample

I run this once and then keep it stored because it takes a long time
to run

```{r}
# Call it all once to set up the structure
# samp = samps[1]
# dr = dirs[1]
# bamf = paste(dr, 'Aligned.sortedByCoord.out.bam', sep = '/')
# fc = call_fc(bamf, gtf, genome)
# stats = (fc$stat
#          %>% data.frame()
#          %>% column_to_rownames('Status'))
# colnames(stats) = samp
# gene_counts[,samp] = fc$counts
# 
# fc_lst = list()
# fc_lst[[samp]] = fc
```

### Run on the rest of the samples

I run this once and then keep it stored because it takes a long time
to run

```{r}
# Call the same thing on all the mapping files

# for (i in 2:length(dirs)){
#     samp = samps[i]
#     dr = dirs[i]
#     bamf = paste(dr, 'Aligned.sortedByCoord.out.bam', sep = '/')
#     fc = call_fc(bamf, gtf, genome)
#     gene_counts[,samp] = fc$counts
#     stats[fc$stat$Status,samp] = fc$stat$Aligned.sortedByCoord.out.bam
#     fc_lst[[samp]] = fc
# }
# 
# # get gene lengths
# gene_lens = select(fc$annotation, GeneID, Length)
# head(gene_lens)
# rownames(gene_counts) = rownames(fc$counts)
# 
# save(list = c('gene_counts', 'stats', 'fc_lst', 'gene_lens'),
#      file = 'results/06_analysis/feature_counts.RData')
load('results/06_analysis/feature_counts.RData')
# head(gene_counts)
```

## Differential Expression

The details of this will depend on your scientific question, but here are the
basic functions you'll almost always need:

```{r, warning=FALSE, message=FALSE}
samdat
# head(gene_counts)

# Remove all rows with only zeros
gene_ct_nz = gene_counts[rowSums(gene_counts) > 0,]
dim(gene_ct_nz)
dds = DESeqDataSetFromMatrix(gene_ct_nz, samdat, ~design_formula)
dds = DESeq(dds)
```

```{r}
res = lfcShrink(dds, type = 'apeglm', contrast = ctrst)
res_df = data.frame(res)

```

#### Plots

Volcano plots for each pairwise comparison

```{r}
ggplot(res_df, aes(x = log2FoldChange, y = -log(padj))) +
    geom_point() +
    facet_wrap(~contrast)
```
PCA plot on variance-stabilized data

```{r}
DESeq2::plotPCA(varianceStabilizingTransformation(dds),intgroup = 'Group') +
    scale_colour_brewer(palette = 'Dark2')
```
